### PR TITLE
Explore: deprecate compact URL encoding & dafault to extended

### PR DESCRIPTION
**Pull request by Elfo404**
**What this PR does / why we need it**:
Removes compact query string encoding in explore and defaults to the extended encoding.

Currently, Explore encodes pane states in arrays in the `left` and `right` parameters.
Items in this array have different meanings based on their position:
```
[timeFrom, timeTo, datasource, ...queries[, extraParameters]]
```

This is problematic when injecting `extraParameters` as, in order to check whether an object is a query or an "extra parameter", we need to check for the presence of specific keys and it's not very scalable. 
For example in https://github.com/grafana/grafana/pull/44017 we introduced a `panelsState` object in the query string, and because of the logic mentioned above, to avoid potential conflicts with queries that may have the same key in their encoded representation, we encode this parameter as `__panelsState`. see https://github.com/grafana/grafana/pull/44017/files#diff-ccef2c5db041700ad4f35f71fdb30d59d9d23f6d02c4877704cf7687eb1a5b8aR253 and https://github.com/grafana/grafana/pull/44017/files#diff-c1f9f4853235e6df01609aa6f51331a77beaad49fb3ba578e5a523390bd927ecR209-R214.
While the solution is ok, other datasource plugins may have queries with the same key, making said datasources unusable I explore.

This PR deprecates the compact encoding by always using the extended version (a normal JSON representation of the state) when generating Explore URLs so that properties are explicitly keyed, for example:
```
/explore?left={"datasource":"gdev-testdata","queries":[{"refId":"A"}],"range":{"from":"now-1h","to":"now"}}
```

the compact encoding is still supported when accessing explore via a manually inputted URL in order to maintain compatibility with URLs generated prior to this change, but it won't be used to generate new URLs.

**How to test**:
Run a previous grafana version that generates compact explore URLs:
```
devenv sources=grafana grafana_version=8.3.4
```
and play a bit with Explore.
Then copy URLs generated from the container and paste them in the version from this branch and it should open the same Explore pages.

(opening an URL with a datasource not configured in one of the instance should fallback to the current logic of selecting a datasource in explore (`last used datasource || default datasource` ))

When pasting URLs in the new version, the resulting URL in the browser should be transformed in the non-compact version
(ie. using `/explore?left=["now-1h","now","Tempo",{"refId":"A"}]` should result in `/explore?left={"datasource":"gdev-testdata","queries":[{"refId":"A"}],"range":{"from":"now-1h","to":"now"}}`

Also, URLs should work both ways, so also URLs generated from this branch should work in 8.3.4

**Other notes**:
- In query history only queries are saved, every other parameter is ignored. we may want to consider changing this behavior in the future
- shortened links save the whole URL so they should work with other parameters as well. to test them you can add `,"panelsState":{"trace":{"spanId":"1"}}` inside the `left` or `right` query string parameters (this was introduced in https://github.com/grafana/grafana/pull/44017).

### CODE REVIEW

File: public/app/core/utils/explore.test.ts [hunk 1]
```diff
@@ -114,6 +114,7 @@ describe('state functions', () => {
       );
     });
 
+    // TODO: remove in 9.0
```
Comment (ifrost) [inline public/app/core/utils/explore.test.ts:117]: Could you add a task to remove it so it's not forgotten?

File: packages/grafana-data/src/utils/url.ts [hunk 2]
```diff
@@ -199,19 +199,12 @@ export const urlUtil = {
  * Create an string that is used in URL to represent the Explore state. This is basically just a stringified json
  * that is that used as a state of a single Explore pane so it does not represent full Explore URL.
  *
- * There are 2 versions of this, normal and compact. Normal is just the same object stringified while compact turns
- * properties of the object into array where the order is significant.
  * @param urlState
- * @param compact
+ * @param compact this parameter is deprecated and will be removed in a future release.
```
Comment (ifrost) [inline packages/grafana-data/src/utils/url.ts:203]: What do we need to do before we remove compact mode? Migrate short links?

**Review by ifrost — APPROVED** [review]
Tested with single pane and split view and seems to be working as expected 👍

**Review by connorlindsey — APPROVED** [review]
Looks great!

I tested 1 and 2 panes with and without a focused span, so all 4 cases worked.

File: packages/grafana-data/src/utils/url.ts (hunk 2 above)
Comment (Elfo404) [reply packages/grafana-data/src/utils/url.ts:203]: nothing actually, the parsing logic will still work with compact URLs. when we then decide to completely remove support for compact URLs we only need to make sure explore doesn't break.

**Pull request merged**